docs/contributing.md

```diff
@@ -0,0 +1,174 @@
+  * Only member functions and no data fields allowed inside `public` scope in public APIs
+  * Two-space indent for other class statements
+  * Leading colon between constructor and member initialization. No space/indent for constructor initializer list. Trailing commas between members.
+  * Do not use `struct` in public APIs. Usage in internal implementation is allowed.
+  * Abstract interface classes should contain only pure abstract member functions. No data fields or function implementations allowed
+
+
+
+**Namespaces**
+* Cuddle brace for namespaces with a space between the name and the opening brace
+* No indentation for namespace contents (including nested namespaces)
+* Closing brace to be followed with a comment with the namespace name, e.g,
+```
+namespace A {
+
+/* code */
+
+} // namespace A
+```
+
+**Pointer and Reference Syntax Alignment**
+  * Left-align `*` and `&`
+
+**Comments and Doc Comments**
+  * Use /// and /** */ comments for documentation purposes and // style comments for notes and general comments
+
+**Examples**
+
+Description of a class
+```
+namespace A {
+namespace B {
+class UpperCamelCase
+{
+public:
+
+  UpperCamelCase(Foo foo, Bar bar)
+  : _foo(foo),
+    _bar(bar),
+    _baz(Baz(foo, bar))
+  {
+  } 
+
+  /// This is an example function
+  ///
+  /// And here is a longer description blah blah blah
+  ///
+  /// \param[in] in_value
+  ///   It takes in a value
+  ///
+  /// \param[out] out_value
+  ///   It puts out a value
+  ///
+  /// \return some result
+  ResultType snake_case_member_functions(
+      InputValue in_value,
+      OutputValue& out_value) const;
+
+private:
+  /* ... etc ... */
+};
+
+} // namespace B
+} // namespace A
+
+```
+
+A class that is only used internally
+
+```
+class ImplementationClass
+{
+public:
+
+  /// documentation
+  double snake_case_public_members_uncommon;
+
+  void foo(Bar bar);
+
+private:
+
+  int _lead_with_underscore;
+
+};
+```
+
+A class that defines an interface
+
+```
+
+class AbstractInterfaceClass
+{
+public:
+  
+  /// Only pure abstract member functions.
+  /// No data fields or implemented
+  ///functions
+  virtual ReturnType pure_virtual_function(
+     SomeArgType arg1,
+     SomeArgType arg2) = 0;
+
+};
+```
+
+**RMF Linter**
+
+Most of these styles and restrictions can be checked with the `ament_uncrustify` command line tool using this [configuration](https://github.com/osrf/rmf_core/blob/master/rmf_utils/test/format/rmf_code_style.cfg) file.
+
+Example usage:
+```
+cd workspace/
+wget https://raw.githubusercontent.com/osrf/rmf_core/master/rmf_utils/test/format/rmf_code_style.cfg
+source /opt/ros/foxy/setup.bash
+ament_uncrustify -c emf_code_style.cfg .
+```
+The `--reformat` option may be passed into the `ament_uncrustify` call to apply the changes in place. However, this is recommended only after manually reviewing the changes.
+
+### Python
+### Version
+
+Python 3 is targeted for development.
```

Comment:
Do we need to specify a minor version? Many library features are not available in earlier minor versions.

Ubuntu Focal has Python 3.8 as the default.

Reply:
That is a good point. Should we specify Python 3.6.5 and above?

Reply:
@mxgrey What minimum Python features are required?

Reply:
If this document is specific to `rmf_core`, then we're not using Python at all in this repo, and I don't ever intend to. The Python bindings for the fleet adapter are in a [separate repository](https://github.com/osrf/rmf_fleet_adapter_python), so I would put any declarations about Python requirements in there.

Reply:
The document is not specific to `rmf_core` and is meant to serve as a guide for all repos under the RMF umbrella.

Reply:
I'm not really inclined to declare a requirement without a specific reason.

`rmf_core` doesn't use Python, and it doesn't put any requirement on any RMF user to use any particular version of Python. Someone could choose to create a fleet adapter using Python 2, and it would be no sweat off our backs.

`traffic-editor` uses Python, so that could be a reference point for what version of Python to require, but `traffic-editor` is not a tool that RMF users are required to use. More importantly, `traffic-editor` is an offline tool; it's not part of an operational deployment of RMF. An RMF user could install Python 3.6.9 on the computer where they use `traffic-editor` but then use Python 3.5 for their fleet adapter, and there would be no inherent problem with that.

My opinion would be that we just shouldn't mention Python here, and that this contributions document should not govern all of RMF, much like the ROS2 contribution guidelines don't govern the entire ROS2 ecosystem (they just govern the core modules that are developed at OSRF).

Reply:
Thanks for the clarification. I've removed the python section from the doc.